### PR TITLE
Tx manager

### PR DESCRIPTION
**Pull request by pablodeymo**
Transaction manager, includes the gas estimation of the transaction.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -66,6 +66,7 @@ bn254 = { git = "https://github.com/sedaprotocol/bn254" }
 eigen-metrics = { version = "0.0.1-alpha", path = "crates/metrics/" }
 serde = { version = "1.0.197", features = ["derive"] }
 eigen-chainio-utils = { path = "crates/chainio/utils/" }
+eigen-chainio-txmanager = { path = "crates/chainio/txmanager/" }
```
Comment (dralves) [inline Cargo.toml:69]: alphabetical order, please, here and elsewhere

File: crates/chainio/txmanager/src/simple_tx_manager.rs [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    /// # Errors
+    ///
+    /// - If the URL is invalid.
+    pub fn new(
+        logger: &'log TracingLogger,
+        gas_limit_multiplier: f64,
+        private_key: &str,
+        rpc_url: &str,
+    ) -> Result<SimpleTxManager<'log>, TxManagerError> {
+        let url = Url::parse(rpc_url).map_err(|_| TxManagerError::InvalidUrlError)?;
```
Comment (dralves) [inline crates/chainio/txmanager/src/simple_tx_manager.rs:60]: can you add some logging here and elsewhere? errors in particular, but successes too if the go sdk also does it

File: crates/chainio/txmanager/src/simple_tx_manager.rs [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    pub async fn wait_for_receipt(
+        pending_tx: PendingTransactionBuilder<'_, Transport, Ethereum>,
+    ) -> Result<TransactionReceipt, TxManagerError> {
+        pending_tx
+            .get_receipt()
+            .await
+            .map_err(|_| TxManagerError::WaitForReceiptError)
+    }
+
+    /*
```
Comment (dralves) [inline crates/chainio/txmanager/src/simple_tx_manager.rs:198]: unused or missing?

File: crates/chainio/txmanager/src/simple_tx_manager.rs [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    use alloy_primitives::{bytes, TxKind::Call, U256};
+    use eigen_logging::{log_level::LogLevel, logger::Logger, tracing_logger::TracingLogger};
+    use once_cell::sync::OnceCell;
+    use tokio;
+
+    static TEST_LOGGER: OnceCell<TracingLogger> = OnceCell::new();
+    const PRIVATE_KEY: &str = "dcf2cbdd171a21c480aa7f53d77f31bb102282b3ff099c78e3118b37348c72f7";
+
+    #[tokio::test]
+    async fn test_send_signed_transaction() {
```
Comment (dralves) [inline crates/chainio/txmanager/src/simple_tx_manager.rs:229]: can you test the legacy and send_eip1559_tx separately

**Review by samlaf — COMMENTED** [review]
I haven't played with rust, ethers-rs, or alloy nearly enough to know what's the best way to do this.
But I'm just wondering in general whether the tx_manager here is necessary at all in a rust-sdk.
For example in the go-sdk I'm implementing a gas-bumping tx_manager in https://github.com/Layr-Labs/eigensdk-go/pull/284. But I think this is not needed in rust because of ethers-rs' middlewares right? https://www.gakonst.com/ethers-rs/middleware/middleware.html

You guys have more context then me, but I just want to raise the question of whether 1:1 feature parity always makes sense. Is this txmanager abstraction even needed, or could rust clients use use the ethers-rs/allow middlewares directly?

File: Cargo.toml (hunk 1 above)
Comment (pablodeymo) [reply Cargo.toml:58]: done, thanks!

File: crates/chainio/txmanager/src/simple_tx_manager.rs (hunk 4 above)
Comment (pablodeymo) [reply crates/chainio/txmanager/src/simple_tx_manager.rs]: test added

**Comment by pablodeymo** [comment]
@dralves this PR is ready for a new review

File: crates/chainio/txmanager/src/simple_tx_manager.rs (hunk 2 above)
Comment (pablodeymo) [reply crates/chainio/txmanager/src/simple_tx_manager.rs]: this is done

**Review by dralves — APPROVED** [review]

File: crates/chainio/txmanager/src/simple_tx_manager.rs [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+        let config = Config::PrivateKey(self.private_key.clone());
+        Config::signer_from_config(config)
+            .inspect_err(|err| self.logger.error("Failed to create signer", &[err]))
+            .map_err(|_| TxManagerError::SignerError)
+    }
+
+    /// Send is used to send a transaction to the Ethereum node. It takes an unsigned/signed transaction
+    /// and then sends it to the Ethereum node.
+    /// If you pass in a signed transaction it will ignore the signature
+    /// and resign the transaction after adding the nonce and gas limit.
```
Comment (dralves) [inline crates/chainio/txmanager/src/simple_tx_manager.rs:104]:
```suggestion
    /// and re-sign the transaction after adding the nonce and gas limit.
```

File: crates/chainio/txmanager/src/simple_tx_manager.rs [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+
+    fn create_local_signer(&self) -> Result<PrivateKeySigner, TxManagerError> {
+        let config = Config::PrivateKey(self.private_key.clone());
+        Config::signer_from_config(config)
+            .inspect_err(|err| self.logger.error("Failed to create signer", &[err]))
+            .map_err(|_| TxManagerError::SignerError)
+    }
+
+    /// Send is used to send a transaction to the Ethereum node. It takes an unsigned/signed transaction
+    /// and then sends it to the Ethereum node.
```
Comment (dralves) [inline crates/chainio/txmanager/src/simple_tx_manager.rs:102]:
```suggestion
    /// ,sends it to the Ethereum node and waits for the receipt.
```

**Review by dralves — APPROVED** [review]
only minor nits, approving assuming they are resolved

File: crates/chainio/txmanager/src/simple_tx_manager.rs (hunk 6 above)
Comment (juanbono) [reply crates/chainio/txmanager/src/simple_tx_manager.rs]: Done!

File: crates/chainio/txmanager/src/simple_tx_manager.rs (hunk 5 above)
Comment (juanbono) [reply crates/chainio/txmanager/src/simple_tx_manager.rs]: Done!

**Comment by juanbono** [comment]
Comments addressed @dralves @supernovahs

**Pull request merged**